app/javascript/pages/country/widget/widgets/widget-fao-forest/widget-fao-forest-selectors.js

```diff
@@ -0,0 +1,78 @@
+import { createSelector } from 'reselect';
+import { format } from 'd3-format';
+
+// get list data
+const getFAO = state => state.fao || null;
+const getRank = state => state.rank || null;
+const getLocationNames = state => state.locationNames || null;
+
+// get lists selected
+export const getFAOForestData = createSelector(
+  [getFAO, getRank, getLocationNames],
+  (fao, rank, locationNames) => {
+    if (!fao || !rank || !locationNames) return null;
+
+    const {
+      area_ha,
+      extent,
+      forest_planted,
+      forest_primary,
+      forest_regenerated
+    } = fao;
+
+    const naturallyRegenerated = extent / 100 * forest_regenerated;
+    const primaryForest = extent / 100 * forest_primary;
+    const plantedForest = extent / 100 * forest_planted;
+    const nonForest =
+      area_ha - (naturallyRegenerated + primaryForest + plantedForest);
+    const total =
+      naturallyRegenerated + primaryForest + plantedForest + area_ha;
+    return {
+      chartData: [
+        {
+          name: 'Naturally regenerated Forest',
+          value: naturallyRegenerated,
+          percentage: naturallyRegenerated / total * 100,
+          color: '#959a00',
+          legendValue: format('.1f')(naturallyRegenerated / total * 100)
```

Comment:
I think I understand now. Good idea about the legend value. Can we avoid duplicating the key props, is legendValue and percentage both needed? I am not sure data selectors are the right place to be putting formatting?

Reply:
I think both are necessary because maybe you want format the legend data completely different as the tooltip, but you're right, in this case the values and the format are the same, so, maybe it is over-engineering if we never have the situation that I had thought.